controllers/clustermesh/clustermesh_controller.go

```diff
@@ -74,117 +86,217 @@ type ClusterMeshReconciler struct {
 //+kubebuilder:rbac:groups=clustermesh.infrastructure.wildlife.io,resources=clustermeshes/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=clustermesh.infrastructure.wildlife.io,resources=clustermeshes/finalizers,verbs=update
 
-func (r *ClusterMeshReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, rerr error) {
-	start := time.Now()
-	r.log = ctrl.LoggerFrom(ctx)
+func DefaultReconciler(mgr manager.Manager) *ClusterMeshReconciler {
+	r := &ClusterMeshReconciler{
+		Client:                         mgr.GetClient(),
+		Scheme:                         mgr.GetScheme(),
+		NewEC2ClientFactory:            ec2.NewEC2Client,
+		PopulateClusterSpecFactory:     PopulateClusterSpec,
+		ReconcilePeeringsFactory:       ReconcilePeerings,
+		ReconcileSecurityGroupsFactory: ReconcileSecurityGroups,
+		ReconcileRoutesFactory:         ReconcileRoutes,
+	}
+
+	return r
+}
+
+func awsConfigForCredential(ctx context.Context, region string, accessKey string, secretAccessKey string) (aws.Config, error) {
+	return config.LoadDefaultConfig(ctx,
+		config.WithRegion(region),
+		config.WithCredentialsProvider(credentials.StaticCredentialsProvider{
+			Value: aws.Credentials{
+				AccessKeyID:     accessKey,
+				SecretAccessKey: secretAccessKey,
+			},
+		}),
+	)
+}
+
+func (c *ClusterMeshReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r := &ClusterMeshReconciliation{
+		ClusterMeshReconciler: *c,
+		start:                 time.Now(),
+		log:                   ctrl.LoggerFrom(ctx),
+	}
 
 	cluster := &clusterv1beta1.Cluster{}
-	if err := r.Get(ctx, req.NamespacedName, cluster); err != nil {
+	if err := c.Get(ctx, req.NamespacedName, cluster); err != nil {
 		return resultError, err
 	}
 
-	clustermesh := &clustermeshv1beta1.ClusterMesh{}
+	if cluster.Spec.ControlPlaneRef == nil || cluster.Spec.ControlPlaneRef.Kind != "KopsControlPlane" {
+		return resultDefault, fmt.Errorf("controlPlaneRef not specified or not supported")
+	}
+
+	r.kcp = &kcontrolplanev1alpha1.KopsControlPlane{}
+	key := client.ObjectKey{
+		Namespace: cluster.Spec.ControlPlaneRef.Namespace,
+		Name:      cluster.Spec.ControlPlaneRef.Name,
+	}
+	if err := c.Get(ctx, key, r.kcp); err != nil {
+		return requeue1min, err
+	}
+	subnet, err := kops.GetSubnetFromKopsControlPlane(r.kcp)
+	if err != nil {
+		r.log.Error(err, fmt.Sprintf("failed to get subnet from kcp"))
+		return requeue1min, err
+	}
+
+	region, err := kops.GetRegionFromKopsSubnet(*subnet)
+	if err != nil {
+		r.log.Error(err, fmt.Sprintf("failed to get region from kcp"))
+		return requeue1min, err
+	}
+
+	var secretName, namespace string
+	if r.kcp.Spec.IdentityRef != nil {
+		secretName = r.kcp.Spec.IdentityRef.Name
+		// gambiarra??
```

Comment:
Remove this comment

Reply:
or elaborate more about why we need this workaround